pkg/blockbuilder/blockbuilder.go

```diff
@@ -528,28 +532,27 @@ func (b *BlockBuilder) consumePartition(
 			lag--
 
 			recProcessedBefore := rec.Timestamp.UnixMilli() <= seenTillTs
-			allSamplesProcessed, err := builder.process(ctx, rec, lastBlockMax, blockMax, recProcessedBefore)
+			allSamplesProcessed, err := builder.process(ctx, rec, lastBlockEnd, blockEnd, recProcessedBefore)
 			if err != nil {
 				// TODO(codesome): do we just ignore this? What if it was Mimir's issue and this leading to data loss?
 				level.Error(b.logger).Log("msg", "failed to process record", "part", part, "key", string(rec.Key), "err", err)
 				continue
 			}
 			if !allSamplesProcessed && commitRec == nil {
-				// If block builder restarts, it will start consuming from the record after this from kafka.
+				// If block builder restarts, it will start consuming from the record after this from kafka (from the commit record).
 				// So the commit record should be the last record that was fully processed and not the
 				// first record that was not fully processed.
 				commitRec = lastRec
 
-				// The first record itself was not fully processed. Meaning the record before
+				// The first record itself was not fully processed, meaning the record before
 				// this is the commit point.
 				if commitRec == nil {
-					// When the commit record does not change from the previous commit, we get nil here.
 					commitRec = &kgo.Record{
 						Timestamp:   time.UnixMilli(pl.CommitRecTs),
 						Topic:       lastCommit.Topic,
 						Partition:   lastCommit.Partition,
 						LeaderEpoch: lastCommit.LeaderEpoch,
-						Offset:      lastCommit.At - 1,
+						Offset:      lastCommit.At - 1, // TODO(v): lastCommit.At can we zero if there wasn't any commit yet
```

Comment:
On this one

> `lastCommit.At` can we zero if there wasn't any commit yet

I'm not sure if there are other cases, where this can happen. We will need to investigate this separately.